File: packages/types/src/calendar.ts
// packages/types/src/calendar.ts
//
// Calendar Intelligence contracts — Phase 5 P0.
// Fuses muhurat + festival + school + govt calendars. Muhurat dates are
// deterministic (cross-checked vs Drik Panchang); this contract is the storage
// shape only — the CalendarOracle service (later tier) reads/overlays them.

export const CalendarEventKind = {
  MUHURAT:  'MUHURAT',
  FESTIVAL: 'FESTIVAL',
  SCHOOL:   'SCHOOL',    // school-holiday windows (affect guest availability)
  GOVT:     'GOVT',      // public / govt holidays
  REGIONAL: 'REGIONAL',
  BLACKOUT: 'BLACKOUT',  // inauspicious window (e.g. Chaturmas)
} as const
export type CalendarEventKind = typeof CalendarEventKind[keyof typeof CalendarEventKind]

export const AuspiciousBand = {
  NONE:   'NONE',
  LOW:    'LOW',
  MEDIUM: 'MEDIUM',
  HIGH:   'HIGH',
  PEAK:   'PEAK',
} as const
export type AuspiciousBand = typeof AuspiciousBand[keyof typeof AuspiciousBand]

/** A single calendar overlay entry (one date or date-range). */
export interface CalendarEvent {
  id:             string
  kind:           CalendarEventKind
  name:           string
  eventDate:      string                          // YYYY-MM-DD (inclusive start)
  endDate:        string | null                   // YYYY-MM-DD, null = single day
  region:         string | null                   // null = national
  source:         string                          // 'DrikPanchang' | 'GOI' | ...
  auspiciousBand: AuspiciousBand
  metadata:       Record<string, unknown> | null  // tithi/nakshatra etc.
  createdAt:      string
}

File: apps/ai-service/src/schemas/calendar.py
"""
Calendar Intelligence schemas (Phase 5 Tier 1).

Pydantic models for the deterministic muhurat / festival calendar.
Mirrors packages/schemas/src/calendar.ts (kind + auspicious_band enums).
No ML, no LLM — pure data + math.
"""

from __future__ import annotations

from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field

CalendarEventKind = Literal["MUHURAT", "FESTIVAL", "SCHOOL", "GOVT", "REGIONAL", "BLACKOUT"]
AuspiciousBand = Literal["NONE", "LOW", "MEDIUM", "HIGH", "PEAK"]


class CalendarEvent(BaseModel):
    """A single calendar overlay row (muhurat / festival / govt holiday)."""

    kind: CalendarEventKind
    name: str
    event_date: str = Field(description="ISO date YYYY-MM-DD")
    end_date: Optional[str] = None
    region: Optional[str] = None  # None = national
    source: str
    auspicious_band: AuspiciousBand = "NONE"
    metadata: Optional[Dict[str, str]] = None  # tithi / nakshatra for muhurats


class MuhuratEvent(BaseModel):
    """A vivah (marriage) muhurat date with panchang metadata."""

    date: str = Field(description="ISO date YYYY-MM-DD")
    band: AuspiciousBand
    # Optional: convention-promoted disputed dates (e.g. the Jan Kharmas dates)
    # may not yet have sourced tithi/nakshatra. Live dataset rows always carry them.
    tithi: Optional[str] = None
    nakshatra: Optional[str] = None


class ChaturmasWindow(BaseModel):
    """
    The Chaturmas blackout for a year. Weddings are blocked in
    [devshayani, devuthani) — weddings resume on the devuthani date.
    """

    year: int
    devshayani: str  # Ashadha Shukla Ekadashi — Chaturmas begins
    devuthani: str  # Kartika Shukla (Prabodhini) Ekadashi — Chaturmas ends


class MuhuratListResponse(BaseModel):
    year: int
    count: int
    chaturmas: ChaturmasWindow
    muhurats: List[MuhuratEvent]


class CalendarEventsResponse(BaseModel):
    count: int
    events: List[CalendarEvent]
